backend/review/views.py

```diff
@@ -1,4 +1,201 @@
-# from django.shortcuts import render
+from django.db.models import F, OuterRef, Q
+from django.http import Http404
+from django.shortcuts import get_object_or_404
+from django.urls import reverse
+from options.models import get_value
+from rest_framework.decorators import api_view
+from rest_framework.response import Response
 
+from courses.models import Course, Department, Instructor
+from review.annotations import annotate_average_and_recent, review_averages
+from review.models import ALL_FIELD_SLUGS, Review
 
-# Create your views here.
+
+def to_r_camel(s):
+    """
+    Turns fields from python snake_case to the PCR frontend's rCamelCase.
+    """
+    return "r" + "".join([x.title() for x in s.split("_")])
+
+
+def make_subdict(field_prefix, d):
+    """
+    Rows in a queryset that don't represent related database models are flat. But we want
+    our JSON to have a nested structure that makes more sense to the client. This function
+    takes fields from a flat dictionary with a certain prefix
+    """
+    start_at = len(field_prefix)
+    return {
+        to_r_camel(k[start_at:]): v
+        for k, v in d.items()
+        if k.startswith(field_prefix) and v is not None
+    }
+
+
+def make_review_response(top_level, nested, nested_name, nested_key):
+    """
+    Instructor and course responses are formatted exactly the same. Factor out the
+    response to remove duplicated code.
+    """
+    return Response(
+        {
+            "average_reviews": make_subdict("average_", top_level),
+            "recent_reviews": make_subdict("recent_", top_level),
+            "num_semesters": top_level["average_semester_count"],
+            nested_name: {
+                r[nested_key]: {
+                    nested_key: r[nested_key],
+                    "average_reviews": make_subdict("average_", r),
+                    "recent_reviews": make_subdict("recent_", r),
+                    "latest_semester": r["recent_semester_calc"],
+                    "num_semesters": r["average_semester_count"],
+                }
+                for r in nested
+            },
+        }
+    )
+
+
+@api_view(["GET"])
+def course_reviews(request, course_code):
```

Comment:
This is more of a question than a comment -- would there be any benefit to eventually using a viewset for these routes?  How come we don't already?

Reply:
ViewSets define REST actions that make a lot of sense when you're talking about a specific "resource" -- generally in our cases, these resources are individual django models that have serializers defined for them. With these aggregations, though, I couldn't find a way to fit the desired output format into a serializer that made much sense. There probably is a way to do it, but I felt like it'd be shoe-horned in so much that it made more sense to use "bare" ApiViews.